danny-avila/LibreChat
# 🧠 feat: Prompt caching switch, prompt query params; refactor: static cache, prompt/markdown styling, trim copied code, switch new chat to convo URL

## Summary

- Implemented a UI switch for Anthropic prompt caching, allowing users to toggle this feature on and off.
- Added support for query parameters to pre-fill the chat input, improving integration possibilities.
- Refactored the role update system to handle multiple permission types simultaneously, enhancing flexibility and reducing database operations.
- Updated the static cache configuration to use a more reasonable one-day time frame for better performance and freshness balance.
- Improved styling for prompt cards and nested lists, enhancing readability and visual appeal.
- Fixed issues with streaming response cursor in nested lists and lists with code blocks (regression, https://github.com/danny-avila/LibreChat/pull/3777)
- Copying code now no longer has newlines
- When a new chat is finalized, the conversation URL switches from `/c/new` to `/c/${UUID}`

**More details:**
- Added a new `promptCache` option to the Anthropic client and settings UI.
- Implemented `useQueryParams` hook to handle URL query parameters for pre-filling chat input.
- Refactored `updateAccessPermissions` function to handle multiple permission types in a single operation.
- Updated static cache configuration for `sMaxAge` and `maxAge` to 1 day and 2 days instead of 1 week and 4 weeks respectively.
- Improved styling for prompt cards, including better handling of long text and nested lists.
- Implemented a streaming cursor for better visual feedback during message generation.
- Added unit tests for the new role update system and other components.
- `copy-to-clipboard` includes a newline by default. For code blocks, this was changed to trim the output to avoid including extraneous newlines
- the `finalHandler` now sets the URL according to the conversationId upon successful completion.

## Change Type

**- [ ] TODO: This change requires a documentation update**

https://github.com/LibreChat-AI/librechat.ai/issues/110
https://github.com/LibreChat-AI/librechat.ai/issues/109

## Testing

To test these changes:

1. Start the application and navigate to the Anthropic settings.
2. Verify that the new "Use Prompt Caching" switch is present and functional.
3. Test the query parameter functionality by appending `?prompt=Hello%20World` to the URL and checking if the chat input is pre-filled.
4. Verify that the streaming cursor appears and behaves correctly during message generation.
5. Check the styling of prompt cards and nested lists for improved readability.
6. Copy code, observe no newlines
7. Observe UUID in URL after conversation finalizes

## Checklist
